File: ARRAY/find max sum such that no two elements are adjacent.cpp
// FIND MAX SUM SUCH THAT NO TWO ELEMENTS ARE ADJACENT

// REQUIRED TC: O(N)

// refer notebook for comments

#include<bits/stdc++.h>
using namespace std;

int findMax(int a[],int n)
{
	int incl=a[0],excl=0,excl_temp;

	for(int i=1;i<n;i++)
	{
		excl_temp=max(incl,excl);

		incl=excl+a[i];

		excl=excl_temp;
	}

	return max(incl,excl);

}
int main()
{
	int arr[]={3,2,7,10};
	int n=sizeof(arr)/sizeof(arr[0]);

	int ans=findMax(arr,n);
	cout<<ans<<"\n";
}

File: ARRAY/find max sum such that no two elements are adjacent.java
package apple;

class maxSum{
	
	static int max(int a,int b)
	{
		return (a>b)?a:b;
	}
	
	static int func(int[] arr,int n)
	{
			int incl=arr[0],excl=0,excl_temp;
			for(int i=1;i<n;i++)
			{
				excl_temp=max(incl,excl);
				incl=excl+arr[i];
				excl=excl_temp;
			}
			
		return max(incl,excl);
	}
	
	public static void main(String args[])
	{
			int[] arr={3,2,7,10};
			int n=arr.length;
			
			System.out.println("the max sum is :: "+func(arr,n));
	}
}